src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py

```diff
@@ -100,17 +105,170 @@ def pileupSizeTask(self):
+        - set expiration date (to be 24h) for already existing ruleIds from pileup document
+
+        :return: None
+        """
+        spec = {'active': True}
+        docs = self.mgr.getPileup(spec)
+        for doc in docs:
+            self.logger.info("partialPileupTask process %s", doc)
+            fraction = doc['containerFraction']
+
+            # check previous fraction value and proceed only if has changed
+            previousFraction = 1.0
+            transition = doc.get('transition', [])
+            if transition:
+                previousFraction = transition[-1]['containerFraction']
+            if previousFraction == fraction:
+                # do nothing since there is no transition, i.e. skip this record
+                self.logger.info("Pileup name %s has no container fractions changes", doc['pileupName'])
+                continue
+
+            # get rucio DIDs (datasets, in CMS name blocks) for our pileup document as following:
+            # - when container fraction is decreasing: new container is a subset of customName
+            # - when container fraction is increasing: new container is a superset of customName + subset of pileupName
+
+            # usage of block names defined in this logic:
+            # https://github.com/dmwm/WMCore/pull/11807#pullrequestreview-1786778783
+            customBlocks = self.getPileupBlocks(doc, previousFraction, fraction)
+
+            # create new container DID in Rucio for our custom name
+            self.rucioClient.createContainer(doc['customName'], scope=self.customRucioScope)
+            self.logger.info("Create container %s with scope %s", doc['customName'], self.customRucioScope)
+
+            # call rucio APIs to attach custom blocks to our custom container (DID)
+            newRules = []
+            self.rucioClient.attachDIDs(None, doc['customName'], customBlocks, scope=self.customRucioScope)
+            for rse in doc['expectedRSEs']:
+                # create new rule for custom DID using pileup document rse
+                newRules += self.rucioClient.createReplicationRule(doc['customName'], rse)
+
+            self.logger.info("Custom pileup: %s has the following new rules created: %s for RSEs: %s",
+                             doc['customName'], newRules, doc['expectedRSEs'])
+            # set expiration date (to be 24h) for already existing ruleIds from pileup document
+            for rid in doc['ruleIds']:
+                # set expiration date to be 24h ahead of right now
+                opts = {'lifetime': 24 * 60 * 60}
+                self.rucioClient.updateRule(rid, opts)
+                self.logger.info("Custom pileup %s has new lifetime %s", doc['customName'], opts['lifetime'])
+
+            # update pileup document
+            if newRules:
+                # remove duplicate rules
+                newRules = set(newRules)
+                try:
+                    # update rules
+                    doc['ruleIds'] += newRules
```

Comment:
if we append the new rule ids, I believe the old ones will never vanish from this document(?) If you agree with this observation, then please change it to `                    doc['ruleIds'] = newRules`

Reply:
your observation is correct, but my knowledge is limited in terms if we need to keep around old rule ids, that's why I kept the update instead of re-assignment. I'll take your word on it and make it assignment then.